tools/testbench/topology_ipc4.c

```diff
@@ -0,0 +1,1304 @@
+				   ctx->hdr->payload_size, NULL);
+	if (ret < 0) {
+		fprintf(stderr, "error: loading controls\n");
+		goto out;
+	}
+
+	tb_setup_widget_ipc_msg(comp_info);
+
+	/* TODO: drop tplg_ctl to avoid memory leak. Need to store and handle this
+	 * to support controls.
+	 */
+	free(tplg_ctl);
+	return 0;
+
+out:
+	free(tplg_ctl);
+	free(comp_info->ipc_payload);
+	return ret;
+}
+
+static int tb_register_graph(struct testbench_prm *tp, int count)
+{
+	struct tplg_context *ctx = &tp->tplg;
+	int ret = 0;
+	int i;
+
+	for (i = 0; i < count; i++) {
+		ret = tplg_parse_graph(ctx, &tp->widget_list, &tp->route_list);
+		if (ret < 0)
+			return ret;
+	}
+
+	return ret;
+}
+
+static int tb_parse_pcm(struct testbench_prm *tp, int count)
+{
+	struct tplg_context *ctx = &tp->tplg;
+	int ret, i;
+
+	for (i = 0; i < count; i++) {
+		ret = tplg_parse_pcm(ctx, &tp->widget_list, &tp->pcm_list);
+		if (ret < 0)
+			return ret;
+	}
+
+	return 0;
+}
+
+/*
+ * create a list with all widget info
+ * containing mapping between component names and ids
+ * which will be used for setting up component connections
+ */
+static inline int tb_insert_comp(struct testbench_prm *tp)
+{
+	struct tplg_context *ctx = &tp->tplg;
+	struct tplg_comp_info *comp_info;
+	int comp_id = ctx->comp_id;
+	int ret;
+
+	if (ctx->widget->id == SND_SOC_TPLG_DAPM_SCHEDULER)
+		return 0;
+
+	comp_info = calloc(sizeof(struct tplg_comp_info), 1);
+	if (!comp_info)
+		return -ENOMEM;
+
+	comp_info->name = strdup(ctx->widget->name);
+	if (!comp_info->name) {
+		ret = -ENOMEM;
+		goto err;
+	}
+
+	comp_info->stream_name = strdup(ctx->widget->sname);
+	if (!comp_info->stream_name) {
+		ret = -ENOMEM;
+		goto sname_err;
+	}
+
+	comp_info->id = comp_id;
+	comp_info->type = ctx->widget->id;
+	comp_info->pipeline_id = ctx->pipeline_id;
+	ctx->current_comp_info = comp_info;
+
+	ret = tb_parse_ipc4_comp_tokens(tp, &comp_info->basecfg);
+	if (ret < 0)
+		goto sname_err;
+
+	list_item_append(&comp_info->item, &tp->widget_list);
+
+	printf("debug: loading comp_id %d: widget %s type %d size %d at offset %ld is_pages %d\n",
+	       comp_id, ctx->widget->name, ctx->widget->id, ctx->widget->size,
+	       ctx->tplg_offset, comp_info->basecfg.is_pages);
+
+	return 0;
+
+sname_err:
+	free(comp_info->name);
+
+err:
+	free(comp_info);
+	return ret;
+}
+
+/* load dapm widget */
+static int tb_load_widget(struct testbench_prm *tp)
+{
+	struct tplg_context *ctx = &tp->tplg;
+	int ret = 0;
+
+	/* get next widget */
+	ctx->widget = tplg_get_widget(ctx);
+	ctx->widget_size = ctx->widget->size;
+
+	/* insert widget into mapping */
+	ret = tb_insert_comp(tp);
+	if (ret < 0) {
+		fprintf(stderr, "tb_load_widget: invalid widget index\n");
+		return ret;
+	}
+
+	/* load widget based on type */
+	switch (tp->tplg.widget->id) {
+	/* load pga widget */
+	case SND_SOC_TPLG_DAPM_PGA:
+		if (tb_new_pga(tp) < 0) {
+			fprintf(stderr, "error: load pga\n");
+			ret = -EINVAL;
+			goto exit;
```

Comment:
can we use `return ...` in all of these?

Reply:
ditto, you don't need a go-to if you are not doing any cleanup

Reply:
yep